pkg/gadgets/trace/byob/tracer/tracer.go

```diff
@@ -0,0 +1,328 @@
+	case ebpf.RingBuf:
+		t.ringbufReader, err = ringbuf.NewReader(t.collection.Maps[t.printMap])
+	case ebpf.PerfEventArray:
+		t.perfReader, err = perf.NewReader(t.collection.Maps[t.printMap], gadgets.PerfBufferPages*os.Getpagesize())
+	default:
+		return fmt.Errorf("unsupported BPF map type: %q", m.Type())
+	}
+	if err != nil {
+		return fmt.Errorf("create BPF map reader: %w", err)
+	}
+
+	// Attach programs
+	for progName, p := range t.spec.Programs {
+		if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kprobe/") {
+			l, err := link.Kprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		} else if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kretprobe/") {
+			l, err := link.Kretprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		}
+	}
+
+	return nil
+}
+
+func (t *Tracer) run(gadgetCtx gadgets.GadgetContext) {
+	d := t.decoderFactory()
+	for {
+		var rawSample []byte
+
+		if t.ringbufReader != nil {
+			record, err := t.ringbufReader.Read()
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					// nothing to do, we're done
+					return
+				}
+				gadgetCtx.Logger().Errorf("read ring buffer: %w", err)
+				return
+			}
+			rawSample = record.RawSample
+		} else if t.perfReader != nil {
+			record, err := t.perfReader.Read()
+			if err != nil {
+				if errors.Is(err, perf.ErrClosed) {
+					return
+				}
+				gadgetCtx.Logger().Errorf("read perf ring buffer: %w", err)
+				return
+			}
+
+			if record.LostSamples != 0 {
+				gadgetCtx.Logger().Warnf("lost %d samples", record.LostSamples)
+				continue
+			}
+			rawSample = record.RawSample
+		} else {
+			gadgetCtx.Logger().Error("neither ring buffer or perf ring buffer was found")
+			return
+		}
+
+		if uint32(len(rawSample)) < t.mapSizes[t.printMap] {
+			gadgetCtx.Logger().Errorf("read ring buffer: len(RawSample)=%d!=%d",
+				len(rawSample),
+				t.mapSizes[t.printMap])
+			return
+		}
+
+		// FIXME: DecodeBtfBinary has a bug with non-NULL-terminated strings.
+		// For now, ensure the problem does not happen in ebpf
+
+		result, err := d.DecodeBtfBinary(gadgetCtx.Context(), t.valueStruct, rawSample[:t.mapSizes[t.printMap]])
+		if err != nil {
+			gadgetCtx.Logger().Errorf("decoding btf: %w", err)
+			return
+		}
+		b, err := json.Marshal(result)
+		if err != nil {
+			gadgetCtx.Logger().Errorf("encoding json: %w", err)
+			return
+		}
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type: eventtypes.NORMAL,
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: 0},
+			Payload:       fmt.Sprintf("%+v", string(b)),
+		}
+
+		if mnt_ns_id_str, ok := result["mnt_ns_id"]; ok {
+			if mnt_ns_id, ok := mnt_ns_id_str.(uint64); ok {
+				event.MountNsID = mnt_ns_id
+			}
+		}
```

Comment:
We can calculate the mount ns in an alternative way without having to depend on json. Please check the approach used in https://github.com/inspektor-gadget/inspektor-gadget/blob/49f7d34e310be78ab962f006a5700deafd4c1c75/pkg/gadgets/trace/byob/tracer/tracer.go#L262-L285 & https://github.com/inspektor-gadget/inspektor-gadget/blob/49f7d34e310be78ab962f006a5700deafd4c1c75/pkg/gadgets/trace/byob/tracer/tracer.go#L338-L345

All of this relies on a new type I defined in solo_types.h

```c
// Inode id of a mount namespace. It's used to enrich the event in user space
typedef u64 mnt_ns_id_t;
```